File: tests/algorithms/x/PHP/backtracking/match_word_pattern.php
<?php
ini_set('memory_limit', '-1');
$now_seed = 0;
$now_seeded = false;
$s = getenv('MOCHI_NOW_SEED');
if ($s !== false && $s !== '') {
    $now_seed = intval($s);
    $now_seeded = true;
}
function _now() {
    global $now_seed, $now_seeded;
    if ($now_seeded) {
        $now_seed = ($now_seed * 1664525 + 1013904223) % 2147483647;
        return $now_seed;
    }
    return hrtime(true);
}
function _append($arr, $x) {
    $arr[] = $x;
    return $arr;
}
$__start_mem = memory_get_usage();
$__start = _now();
  function get_value($keys, $values, $key) {
  $i = 0;
  while ($i < count($keys)) {
  if ($keys[$i] == $key) {
  return $values[$i];
}
  $i = $i + 1;
};
  return null;
};
  function contains_value($values, $value) {
  $i = 0;
  while ($i < count($values)) {
  if ($values[$i] == $value) {
  return true;
}
  $i = $i + 1;
};
  return false;
};
  function backtrack($pattern, $input_string, $pi, $si, $keys, $values) {
  if ($pi == strlen($pattern) && $si == strlen($input_string)) {
  return true;
}
  if ($pi == strlen($pattern) || $si == strlen($input_string)) {
  return false;
}
  $ch = substr($pattern, $pi, $pi + 1 - $pi);
  $mapped = get_value($keys, $values, $ch);
  if ($mapped != null) {
  if (substr($input_string, $si, $si + strlen($mapped) - $si) == $mapped) {
  return backtrack($pattern, $input_string, $pi + 1, $si + strlen($mapped), $keys, $values);
};
  return false;
}
  $end = $si + 1;
  while ($end <= strlen($input_string)) {
  $substr = substr($input_string, $si, $end - $si);
  if (contains_value($values, $substr)) {
  $end = $end + 1;
  continue;
}
  $new_keys = _append($keys, $ch);
  $new_values = _append($values, $substr);
  if (backtrack($pattern, $input_string, $pi + 1, $end, $new_keys, $new_values)) {
  return true;
}
  $end = $end + 1;
};
  return false;
};
  function match_word_pattern($pattern, $input_string) {
  $keys = [];
  $values = [];
  return backtrack($pattern, $input_string, 0, 0, $keys, $values);
};
  function main() {
  echo rtrim(json_encode(match_word_pattern('aba', 'GraphTreesGraph'), 1344)), PHP_EOL;
  echo rtrim(json_encode(match_word_pattern('xyx', 'PythonRubyPython'), 1344)), PHP_EOL;
  echo rtrim(json_encode(match_word_pattern('GG', 'PythonJavaPython'), 1344)), PHP_EOL;
};
  main();
$__end = _now();
$__end_mem = memory_get_peak_usage();
$__duration = max(1, intdiv($__end - $__start, 1000));
$__mem_diff = max(0, $__end_mem - $__start_mem);
$__bench = ["duration_us" => $__duration, "memory_bytes" => $__mem_diff, "name" => "main"];
$__j = json_encode($__bench, 128);
$__j = str_replace("    ", "  ", $__j);
echo $__j, PHP_EOL;

File: tests/algorithms/transpiler/FS/backtracking/match_word_pattern.fs
// Generated 2025-08-06 20:48 +0700

exception Break
exception Continue

exception Return
let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let _substring (s:string) (start:int) (finish:int) =
    let len = String.length s
    let mutable st = if start < 0 then len + start else start
    let mutable en = if finish < 0 then len + finish else finish
    if st < 0 then st <- 0
    if st > len then st <- len
    if en > len then en <- len
    if st > en then st <- en
    s.Substring(st, en - st)

let _idx (arr:'a array) (i:int) : 'a =
    if i >= 0 && i < arr.Length then arr.[i] else Unchecked.defaultof<'a>
let rec get_value (keys: string array) (values: string array) (key: string) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable keys = keys
    let mutable values = values
    let mutable key = key
    try
        let mutable i: int = 0
        while i < (Seq.length(keys)) do
            if (_idx keys (i)) = key then
                __ret <- _idx values (i)
                raise Return
            i <- i + 1
        __ret <- unbox<string> null
        raise Return
        __ret
    with
        | Return -> __ret
and contains_value (values: string array) (value: string) =
    let mutable __ret : bool = Unchecked.defaultof<bool>
    let mutable values = values
    let mutable value = value
    try
        let mutable i: int = 0
        while i < (Seq.length(values)) do
            if (_idx values (i)) = value then
                __ret <- true
                raise Return
            i <- i + 1
        __ret <- false
        raise Return
        __ret
    with
        | Return -> __ret
and backtrack (pattern: string) (input_string: string) (pi: int) (si: int) (keys: string array) (values: string array) =
    let mutable __ret : bool = Unchecked.defaultof<bool>
    let mutable pattern = pattern
    let mutable input_string = input_string
    let mutable pi = pi
    let mutable si = si
    let mutable keys = keys
    let mutable values = values
    try
        if (pi = (String.length(pattern))) && (si = (String.length(input_string))) then
            __ret <- true
            raise Return
        if (pi = (String.length(pattern))) || (si = (String.length(input_string))) then
            __ret <- false
            raise Return
        let ch: string = _substring pattern pi (pi + 1)
        let mapped: string = get_value (keys) (values) (ch)
        if mapped <> (unbox<string> null) then
            if (_substring input_string si (si + (String.length(mapped)))) = mapped then
                __ret <- backtrack (pattern) (input_string) (pi + 1) (si + (String.length(mapped))) (keys) (values)
                raise Return
            __ret <- false
            raise Return
        let mutable ``end``: int = si + 1
        try
            while ``end`` <= (String.length(input_string)) do
                try
                    let substr: string = _substring input_string si ``end``
                    if contains_value (values) (substr) then
                        ``end`` <- ``end`` + 1
                        raise Continue
                    let new_keys: string array = Array.append keys [|ch|]
                    let new_values: string array = Array.append values [|substr|]
                    if backtrack (pattern) (input_string) (pi + 1) (``end``) (new_keys) (new_values) then
                        __ret <- true
                        raise Return
                    ``end`` <- ``end`` + 1
                with
                | Continue -> ()
                | Break -> raise Break
        with
        | Break -> ()
        | Continue -> ()
        __ret <- false
        raise Return
        __ret
    with
        | Return -> __ret
and match_word_pattern (pattern: string) (input_string: string) =
    let mutable __ret : bool = Unchecked.defaultof<bool>
    let mutable pattern = pattern
    let mutable input_string = input_string
    try
        let keys: string array = [||]
        let values: string array = [||]
        __ret <- backtrack (pattern) (input_string) (0) (0) (keys) (values)
        raise Return
        __ret
    with
        | Return -> __ret
and main () =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    try
        let __bench_start = _now()
        let __mem_start = System.GC.GetTotalMemory(true)
        printfn "%b" (match_word_pattern ("aba") ("GraphTreesGraph"))
        printfn "%b" (match_word_pattern ("xyx") ("PythonRubyPython"))
        printfn "%b" (match_word_pattern ("GG") ("PythonJavaPython"))
        let __bench_end = _now()
        let __mem_end = System.GC.GetTotalMemory(true)
        printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)

        __ret
    with
        | Return -> __ret
main()
